### PR TITLE
Implement tensor Network model and optimization of Ising 1D Open b.c.

### LINKED ISSUE #7: Implement tensor Network model and optimization of Ising 1D Open b.c.
Implement the tensor network model code and training loop for accurately representing the 1D Ising model with open boundary conditions.

Duplicating considerations added to PR here for visibility and reusability across similar optimization problems in the future:

**Todo:** Analyze generated quantum system data and consider tensor network parameter design.

Data related questions:

1. Negative energies normalizations? Ground state alone vs systems including excited states?
2. Entanglement entropy normalizations? Once again, ground state alone vs excitations?
3. Data augmentations to increase training, validation, test sets? E.g., global phases for states?
4. What does batching look like with this dataset?
5. How does the TN representation generalize to other system datasets?

Tensor Network architecture and optimization related questions:

1. How should we initialize the model weights? Xavier (Linear) initialization? Kaiming (ReLU) initialization? Something else?
2. Optimizer: AdamW default, consider Lion for memory intense trainings (likely not applicable)?
3. Learning Rate Scheduling? Linear? Cosine? Step decay?
4. Learning Rate warm up steps? Anticipate this working well with cosine decay.
5. Dropout of some model weights? How does this fit into the tensor paradigm?
6. Batch size? Consider batch size+LR relationship. PyTorch GPUs vs Jax TPUs? See comment above about batched data
7. Early Stopping necessary? What's the prevalence of overfitting here?
8. Train on various loss functions, and permuted combinations thereof    - Quantum state fidelity(ies) (with importance weighted wrt lower energy states, see *** below)
    - Energy eigenstate(s) (with importance weighted wrt lower energy states, see *** below)
    - Entanglement entropies across different cuts (and necessary symmetry properties thereof)
    - Correlation functions and local observables: short-range and long-range correlations, connected correlators
    - NOTE: intuitively a loss function which combines all of these would - *one would think* - most accurately emulate the true quantum mechanics and quantum information that underpin the system; only time will tell how the system responds to training in practice. Consider how various combinations of these loss functions correspond to quantum state emulation accuracy for datasets with a varying number of excited states
8. Depending on performance, consider hybrid neural tensor network (hNTN) architectures to include intermediate layers known to improve NN optimization (normalization layers, etc.).

Data augmentation and network design considerations:

- Normalize inputs for improved gradients
- Initialize tensor network model weights appropriately, try deep learning standards and any TN-specific derivatives thereof
- Optimize hyperparameters iteratively; consider hyperparameter tuning, e.g. `optuna`
- Gradually increase loss function complexity (considering various permutations thereof) and compare accuracy with number of excited states
- Comparison between canonical tensor networks and hNTNs. Points of comparison include: state representation accuracy, training efficiency, hardware compatibility differences (GPUs v TPUs)?  

From *** above: An existing issue on the notion of descending Hilbert space state and energy eigenstate state importance to emulate necessary physics: https://github.com/johngrahamreynolds/holographic_tensor_networks/issues/1

### PR DESCRIPTION
**Pull request by johngrahamreynolds**
See the issue https://github.com/johngrahamreynolds/holographic_tensor_networks/issues/7

### CODE REVIEW

**Comment by johngrahamreynolds** [comment]
**Todo:** Analyze generated quantum system data and consider tensor network parameter design.

Data related questions:

1. Negative energies normalizations? Ground state alone vs systems including excited states?
2. Entanglement entropy normalizations? Once again, ground state alone vs excitations?
3. Data augmentations to increase training, validation, test sets? E.g., global phases for states?
4. What does batching look like with this dataset?
5. How does the TN representation generalize to other system datasets?

Tensor Network architecture and optimization related questions:

1. How should we initialize the model weights? Xavier (Linear) initialization? Kaiming (ReLU) initialization? Something else?
2. Optimizer: AdamW default, consider Lion for memory intense trainings (likely not applicable)?
3. Learning Rate Scheduling? Linear? Cosine? Step decay?
4. Learning Rate warm up steps? Anticipate this working well with cosine decay.
5. Dropout of some model weights? How does this fit into the tensor paradigm?
6. Batch size? Consider batch size+LR relationship. PyTorch GPUs vs Jax TPUs? See comment above about batched data
7. Early Stopping necessary? What's the prevalence of overfitting here?
8. Train on various loss functions, and permuted combinations thereof
    - Quantum state fidelity(ies) (with importance weighted wrt lower energy states, see *** below)
    - Energy eigenstate(s) (with importance weighted wrt lower energy states, see *** below)
    - Entanglement entropies across different cuts (and necessary symmetry properties thereof)
    - Correlation functions and local observables: short-range and long-range correlations, connected correlators
    - NOTE: intuitively a loss function which combines all of these would - *one would think* - most accurately emulate the true quantum mechanics and quantum information that underpin the system; only time will tell how the system responds to training in practice. Consider how various combinations of these loss functions correspond to quantum state emulation accuracy for datasets with a varying number of excited states
9. Depending on performance, consider hybrid neural tensor network (hNTN) architectures to include intermediate layers known to improve NN optimization (normalization layers, etc.).


Data augmentation and network design considerations:

- Normalize inputs for improved gradients
- Initialize tensor network model weights appropriately, try deep learning standards and any TN-specific derivatives thereof
- Optimize hyperparameters iteratively; consider hyperparameter tuning, e.g. `optuna`
- Gradually increase loss function complexity (considering various permutations thereof) and compare accuracy with number of excited states
- Comparison between canonical tensor networks and hNTNs. Points of comparison include: state representation accuracy, training efficiency, hardware compatibility differences (GPUs v TPUs)?  




From *** above: An existing issue on the notion of descending Hilbert space state and energy eigenstate state importance to emulate necessary physics: https://github.com/johngrahamreynolds/holographic_tensor_networks/issues/1
